README.md
### 〓 Instructions 〓
1. Please install python3 and run "prepare.bat" before running it for the first time.
    - Download URL: https://www.python.org/downloads/release/python-377/
    - Version 3.7.7 is recommended, otherwise the installation of the library may generate exceptions.
2. Modify the configuration file ".\Pixiv_Collection\config.json".
    - [chrome] the file path of the chrome browser.
3. Download the driver for your chrome version and place it in the current directory.
    - Please make sure the chrome driver version matches the chrome version as chrome has an auto-update problem.
    - Chrome driver URL: https://googlechromelabs.github.io/chrome-for-testing/
    - <2023-10-10> Backup: https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/117.0.5938.149/win64/chromedriver-win64.zip
4. Please log in to your pixiv account in the pop-up browser in a minute.
5. Just wait for the image to download.

### 〓 Command-line parameter 〓
- h/help
    - View command line prompts.
- debug/debug
    - Turn on debug mode. Do not start the simulator.
- sd/start_date
    - Start date.
    - Exmaple: -sd "20231010"
- ed/end_date
    - End date.
    - Example: -ed "20231011"

### 〓 Latest version 2.1 〓
1. Add customized date intervals.
2. Add history of date to avoid repeated crawling.
3. Add history of image to avoid repeated downloading.

history.md
### 〓 Updated version 2.1 〓
1. Add customized date intervals.
2. Add history of date to avoid repeated crawling.
3. Add history of image to avoid repeated downloading.

#### 〓 Updated version 2.0 〓
1. Automation download image with selenium and requests.

#### 〓 Updated version 1.x 〓
1. Automation download image with selenium.

pixiv.py
#-*- encoding: utf-8 -*-
#!/usr/bin/pixiv_venv python3.7
"""
[File]        : pixiv.py
[Time]        : 2023/10/10 18:00:00
[Author]      : InaKyui
[License]     : (C)Copyright 2023, InaKyui
[Version]     : 2.1
[Description] : Class pixiv.
"""

__authors__ = ["InaKyui <https://github.com/InaKyui>"]
__version__ = "Version: 2.1"

import os
import re
import json
import time
import random
import datetime
import requests
from database import Database
from selenium import webdriver

# Disable warnings from requests.
requests.packages.urllib3.disable_warnings()

class Pixiv:
    def __init__(self, database:Database) -> None:
        self.__config = None
        self.__load_config()
        self.headers = {
            "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36",
            "Referer": "https://www.pixiv.net/"
        }
        self.database = database
        self.session = requests.Session()

        daily = {
            "Type": "daily",
            "URL": "https://www.pixiv.net/ranking.php?mode=daily&date={0}&p={1}",
            "Page": 10
        }
        daily_ai = {
            "Type": "daily_ai",
            "URL": "https://www.pixiv.net/ranking.php?mode=daily_ai&date={0}&p={1}",
            "Page": 1
        }
        daily_r18 = {
            "Type": "daily_r18",
            "URL": "https://www.pixiv.net/ranking.php?mode=daily_r18&date={0}&p={1}",
            "Page": 2
        }
        daily_ai_r18 = {
            "Type": "daily_r18_ai",
            "URL": "https://www.pixiv.net/ranking.php?mode=daily_r18_ai&date={0}&p={1}",
            "Page": 1
        }
        self.__task = [daily, daily_ai, daily_r18, daily_ai_r18]

    def __load_config(self) -> None:
        with open(os.path.join(os.getcwd(), "config.json"), "r", encoding="utf8") as fr:
            self.__config = json.load(fr)

    def __get_cookie(self):
        chrome_cmd = "{0} --remote-debugging-port=9222 --user-data-dir={1}".format(self.__config["chrome"], os.path.join(os.getcwd(), "selenium\chrome"))
        os.popen(chrome_cmd)
        options = webdriver.ChromeOptions()
        options.debugger_address = "127.0.0.1:9222"
        browser = webdriver.Chrome(options=options)
        time.sleep(3)
        browser.get("https://www.pixiv.net/ranking.php")
        time.sleep(60)
        cookie_lst = []
        for item in browser.get_cookies():
            cookie_lst.append("{}={}".format(item["name"], item["value"]))
        cookie = ";".join(cookie_lst)
        self.headers["cookie"] = cookie

    def __get_artworks(self, date:str, task:dict) -> list:
        artworks = []
        for page in range(task["Page"]):
            daily_url = task["URL"].format(date, str(page + 1))
            daily_res = self.session.get(url=daily_url, headers=self.headers, verify=False)
            rank_ptn = re.compile("data-attr=\".*?\" data-id=\".*?\"")
            for rp in rank_ptn.findall(daily_res.text):
                artwork = rp.split("data-id=\"")[1].split("\"")[0]
                artworks.append(artwork)
                print("[Get] Artwork: {}".format(artwork))
        return artworks

    def __get_images(self, artworks:list) -> list:
        images = []
        for artwork in artworks:
            image_info = {}
            time.sleep(random.randint(3,5) / 10)
            # Get image url.
            artwork_url = "https://www.pixiv.net/artworks/" + artwork
            art_rsp = self.session.get(url=artwork_url, headers=self.headers, verify=False)
            # art_ptn = re.compile("\"urls\":\S+?}")
            img_ptn = re.compile("\"original\"\S*[0-9]{4}/[0-9]{2}/[0-9]{2}/[0-9]{2}/[0-9]{2}/[0-9]{2}/[0-9]{9}_p[0-9]\.\S{3,4}\"")
            cfs_lst = img_ptn.findall(art_rsp.text)
            # \"pageCount\":[0-9]+,
            # Non-moving images.
            if cfs_lst == []:
                print("[Fail] {}".format(artwork_url))
                continue
            else:
                # TODO Single image -> Multi image.
                image_url = cfs_lst[0].split("\"")[3]
            # Collecting image information.
            image_info["name"] = image_url.split("/")[-1]
            image_info["url"] = image_url
            image_info["artwork"] = artwork
            date_ptn = re.compile("[0-9]{4}\/[0-9]{2}\/[0-9]{2}\/[0-9]{2}\/[0-9]{2}\/[0-9]{2}")
            date_lst = date_ptn.findall(image_url)
            image_info["upload_time"] = date_lst[0].replace("/", "-") if date_lst else ""
            images.append(image_info)
            print("[Get] Image: {}".format(image_info["url"]))
        return images

    def download_image(self, download_path:str, image_info:dict):
        img_rsp = self.session.get(url=image_info["url"], headers=self.headers, verify=False)
        with open(os.path.join(download_path, image_info["url"].split("/")[-1]), "wb") as fw:
            fw.write(img_rsp.content)
        # Supplement and insert image information.
        image_info["download_time"] = datetime.datetime.now()
        self.database.insert_image(image_info)
        time.sleep(random.randint(10, 15) / 10)
        print("[Download] Image: {}".format(image_info["url"]))

    def download(self, start_date, end_date):
        self.__get_cookie()
        print("[Start] Download.")
        # Leaderboard is updated at 12:00 p.m..
        if int(datetime.datetime.now().strftime("%H")) < 12:
            last_date = (datetime.datetime.now() - datetime.timedelta(days=2))
        else:
            last_date = (datetime.datetime.now() - datetime.timedelta(days=1))
        # Whichever is earlier.
        start_date = last_date if last_date < start_date else start_date
        last_date = last_date if last_date < end_date else end_date
        # Closed interval.
        apart_days = last_date.__sub__(start_date).days + 1
        # Download images.
        for i in range(apart_days):
            cur_date = (start_date + datetime.timedelta(days=i)).strftime("%Y%m%d")
            # Determine whether the current date has been downloaded.
            if not self.database.select_date(cur_date):
                for task in self.__task:
                    # Determine whether the download path exists.
                    download_path = os.path.join(os.getcwd(), "images", task["Type"], cur_date)
                    if not os.path.exists(download_path):
                        os.makedirs(download_path)
                    # Date -> [artworks] -> [images].
                    artworks = self.__get_artworks(cur_date, task)
                    images = self.__get_images(artworks)
                    failure_images = []
                    for iu in images:
                        # Determine whether the image has been downloaded.
                        if not self.database.select_image(iu["name"]):
                            try:
                                iu["type"] = task["Type"]
                                self.download_image(download_path, iu)
                            except:
                                failure_images.append(iu)
                    # Second retry for failure images.
                    for fiu in failure_images:
                        try:
                            fiu["type"] = task["Type"]
                            self.download_image(download_path, fiu)
                        except:
                            print("[Fail] {}".format(fiu["url"]))
                self.database.insert_date(cur_date)

database.py
#-*- encoding: utf-8 -*-
#!/usr/bin/pixiv_venv python3.7
"""
[File]        : main.py
[Time]        : 2023/10/10 18:00:00
[Author]      : InaKyui
[License]     : (C)Copyright 2023, InaKyui
[Version]     : 2.1
[Description] : Class database.
"""

__authors__ = ["InaKyui <https://github.com/InaKyui>"]
__version__ = "Version: 2.1"

import sqlite3
import datetime

class Database:
    def __init__(self, db_name:str):
        self.database = sqlite3.connect(db_name)
        self.cursor = self.database.cursor()

    def create_table(self):
        # Images table.
        sql_create = '''CREATE TABLE images
                        (id INTEGER PRIMARY KEY AUTOINCREMENT,
                         name TEXT NOT NULL,
                         url TEXT NOT NULL,
                         type TEXT NOT NULL,
                         arkwork TEXT NOT NULL,
                         upload_time DATETIME,
                         download_time DATETIME);'''
        self.cursor.execute(sql_create)
        # Dates table.
        sql_create = '''CREATE TABLE dates
                        (id INTEGER PRIMARY KEY AUTOINCREMENT,
                         date TEXT NOT NULL,
                         download_time DATETIME);'''
        self.cursor.execute(sql_create)

    def select_image(self, image_name:str) -> list:
        sql_select = '''SELECT id FROM images WHERE name="{}"'''
        rst = self.cursor.execute(sql_select.format(image_name))
        return rst.fetchall()

    def select_date(self, date:str) -> list:
        sql_select = '''SELECT id FROM dates WHERE date="{}"'''
        rst = self.cursor.execute(sql_select.format(date))
        return rst.fetchall()

    def insert_image(self, image_info:dict):
        sql_insert = '''INSERT INTO images (name, url, type, arkwork, create_time, download_time) VALUES (?,?,?,?,?,?)'''
        data = (image_info["name"],
                image_info["url"],
                image_info["type"],
                image_info["arkwork"],
                image_info["upload_time"],
                image_info["download_time"])
        self.cursor.execute(sql_insert, data)
        self.database.commit()

    def insert_date(self, date:str):
        sql_insert = '''INSERT INTO dates (date, download_time) VALUES (?, ?)'''
        self.cursor.execute(sql_insert, (date, datetime.datetime.now()))
        self.database.commit()

Scripts/Activate.ps1
function global:deactivate ([switch]$NonDestructive) {
    # Revert to original values
    if (Test-Path function:_OLD_VIRTUAL_PROMPT) {
        copy-item function:_OLD_VIRTUAL_PROMPT function:prompt
        remove-item function:_OLD_VIRTUAL_PROMPT
    }

    if (Test-Path env:_OLD_VIRTUAL_PYTHONHOME) {
        copy-item env:_OLD_VIRTUAL_PYTHONHOME env:PYTHONHOME
        remove-item env:_OLD_VIRTUAL_PYTHONHOME
    }

    if (Test-Path env:_OLD_VIRTUAL_PATH) {
        copy-item env:_OLD_VIRTUAL_PATH env:PATH
        remove-item env:_OLD_VIRTUAL_PATH
    }

    if (Test-Path env:VIRTUAL_ENV) {
        remove-item env:VIRTUAL_ENV
    }

    if (!$NonDestructive) {
        # Self destruct!
        remove-item function:deactivate
    }
}

deactivate -nondestructive

$env:VIRTUAL_ENV="C:\Python_Script\pixiv_venv"

if (! $env:VIRTUAL_ENV_DISABLE_PROMPT) {
    # Set the prompt to include the env name
    # Make sure _OLD_VIRTUAL_PROMPT is global
    function global:_OLD_VIRTUAL_PROMPT {""}
    copy-item function:prompt function:_OLD_VIRTUAL_PROMPT
    function global:prompt {
        Write-Host -NoNewline -ForegroundColor Green '(pixiv_venv) '
        _OLD_VIRTUAL_PROMPT
    }
}

# Clear PYTHONHOME
if (Test-Path env:PYTHONHOME) {
    copy-item env:PYTHONHOME env:_OLD_VIRTUAL_PYTHONHOME
    remove-item env:PYTHONHOME
}

# Add the venv to the PATH
copy-item env:PATH env:_OLD_VIRTUAL_PATH
$env:PATH = "$env:VIRTUAL_ENV\Scripts;$env:PATH"


main.py
#-*- encoding: utf-8 -*-
#!/usr/bin/pixiv_venv python3.7
"""
[File]        : main.py
[Time]        : 2023/10/10 18:00:00
[Author]      : InaKyui
[License]     : (C)Copyright 2023, InaKyui
[Version]     : 2.1
[Description] : Code entrance.
"""

__authors__ = ["InaKyui <https://github.com/InaKyui>"]
__version__ = "Version: 2.1"

import os
import argparse
from pixiv import Pixiv
from database import Database
from datetime import datetime

def main(args:argparse.Namespace):
    create_flag = os.path.exists(os.path.join(os.getcwd(), "pixiv.db"))
    pixiv_db = Database("pixiv.db")
    # If the database exists there is no need to create table.
    if not create_flag:
        pixiv_db.create_table()

    # Define the start and end date of the download task.
    start_date = datetime.strptime(args.start_date, "%Y%m%d")
    end_date = datetime.strptime(args.end_date, "%Y%m%d")

    # Download images.
    pixiv = Pixiv(pixiv_db)
    pixiv.download(start_date, end_date)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-debug", "--debug", action="store_true", help="Turn on debug mode.")
    parser.add_argument("-sd", "--start_date", type=str, default=datetime.now().strftime("%Y%m%d"), help="Start date.")
    parser.add_argument("-ed", "--end_date", type=str, default=datetime.now().strftime("%Y%m%d"), help="End date.")
    args = parser.parse_args()
    main(args)